www/js/lib/util.js

```diff
@@ -23,57 +23,65 @@
 
     /**
-     * Returns the same String with the first letter in lower-case
-     * @param {String} string
-     * @returns {String}
+     * Generates an array of strings with all possible combinations of first-letter or all-letter case transformations
+     * If caseMatchType is not 'full', then only combinations of first-letter cases for each word are calculated
+     * If caseMatchType is 'full', then all-uppercase combinations of each word are added to the variations array
+     * NB may produce duplicate strings if string begins with punctuation or if it is in a language with no case
+     * @param {String} string The string to be converted
+     * @param {String} caseMatchType ('basic'|'full') The type (complexity) of case variations to calculate 
+     * @return {Array} An array containing strings with all possible combinations of case types
      */
-    function lcFirstLetter(string) {
+    function allCaseFirstLetters(string, caseMatchType) {
         if (string) {
-            if (string.length >= 1) {
-                return string.charAt(0).toLocaleLowerCase() + string.slice(1);
-            } else {
-                return string;
+            var comboArray = [];
+            // Split string into parts beginning with first word letters
+            var strParts = string.match(regExpFindStringParts);
+            // Set the base (binary or ternary) according to the complexity of the search
+            var base = caseMatchType === 'full' ? 3 : 2;
+            // If n = strParts.length, then the number of possible case combinations (numCombos) is base ^ n
+            // For *basic* case calculation: think of numCombos as a binary number of n bits, with each bit representing lcase (0) or ucase (1)
+            // For *full* case calculation: think of numCombos as a tertiary base number, e.g. 000, 111, 222,
+            // with each bit representing all-lowercase (0), First-Letter-Uppercase (1) or ALL-UPPERCASE (2)
+            var numCombos = Math.pow(base, strParts.length);
+            var typeCase, mixedTypeCaseStr, bitmask, caseBit;
+            // Iterate through every possible combination, starting with (base ^ n) - 1 and decreasing; we go from high to low,
+            // because title case (e.g. binary 1111) is more common than all lowercase (0000) so will be found first
+            for (var i = numCombos; i--;) {
+                mixedTypeCaseStr = '';
+                bitmask = 1;
+                for (var j = 0; j < strParts.length; j++) {
+                    // Get modulus of division (this is equivalent to bitwise AND for different bases)
+                    // caseBit will be 0, 1 or 2 (latter only for 'full' case calcualation)
+                    caseBit = ~~(i / bitmask % base);
+                    if (caseBit === 2) {
+                        // All uppercase
+                        typeCase = strParts[j].toLocaleUpperCase();
```

Comment:
According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase , using `toLocaleUpperCase` without passing a Locale uses the "host environment’s current locale".
Could it be an issue if my device is configured with a locale that is different from the one of the ZIM file? If so, we might need to pass the locale found in the ZIM metadata?

Reply:
This isn't a change from the original code, which already used `toLocaleUpperCase`. But I'll investigate.

Reply:
@mossroy So, looking into this, mostly `toLocaleUpperCase` will return the same result as `toUpperCase`, but in "languages like Turkish" it may vary. So the issue would only present itself if a user were accessing a Turkish (e.g.) ZIM on a machine configured for, say, en-GB locale. Then they might get some funny uppercase  Turkish letters based on UTF-8 mappings, and not on standard Turkish fonts. So, we can't resolve this without more info: does Wikipedia use UTF-8 mappings for upper-lower-case on Turkish ZIMs? Or does it use some locale-specific mapping?

I don't think we should be trying to work out locale from the ZIM metadata, because the language codes in the ZIM do not map one-to-one onto JavaScript/HTML (or Windows/Mac/Linux???) locale strings. Possibly for the currently loaded page there *might* be an HTML5 locale string we could scrape, but it's not a universal solution because we don't know that every ZIM article has locale set.

Finally, the user does have a remedy: type a more exact version of search string.

All-in-all, this is, I think, a minor issue. It's always been in Kiwix JS, it is not part of this PR, and a solution would be a separate issue. Is that OK? If you would like the issue resolved, maybe you could open it as a separate issue? Otherwise I'm resolving this for now!

Reply:
I agree